<a href="https://colab.research.google.com/github/rohan2RS/NETFLIX_movie_recommendation_system/blob/main/MovieRecommendationSystem2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
data = pd.read_csv('/content/sample_data/netflix_titles.csv')
data.head()


show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

# Task
Create a hybrid movie recommendation system using the dataset at "/content/sample_data/netflix_titles.csv".

## Data loading and exploration

### Subtask:
Load the data into a pandas DataFrame and perform initial exploration to understand the structure and content of the data.


**Reasoning**:
Display the first and last 5 rows, show column information and data types, generate descriptive statistics, and check for missing values to explore the data.



In [2]:
display(data.head())
display(data.tail())
data.info()
display(data.describe())
display(data.isnull().sum())

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

show_id     type        title         director  \
8802   s8803    Movie       Zodiac    David Fincher   
8803   s8804  TV Show  Zombie Dumb              NaN   
8804   s8805    Movie   Zombieland  Ruben Fleischer   
8805   s8806    Movie         Zoom     Peter Hewitt   
8806   s8807    Movie       Zubaan      Mozez Singh   

                                                   cast        country  \
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

             date_added  release_year rating   duration  \
8802  November 20, 2019          2007      R    158 min   
8803       July 1, 2019          2018  TV-Y7  2 Seasons   
8804   November 1, 2019          2009      R     88 min   
8805   January 11, 2020          2006     PG     88 min   
8806      March 2, 2019          2015  TV-14    111 min   

                                           listed_in  \
8802                  Cult Movies, Dramas, Thrillers   
8803          Kids' TV, Korean TV Shows, TV Comedies   
8804                         Comedies, Horror Movies   
8805              Children & Family Movies, Comedies   
8806  Dramas, International Movies, Music & Musicals   

                                            description  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

## Data preprocessing

### Subtask:
Handle missing values, convert data types, and perform any necessary feature engineering (e.g., extracting relevant information from text fields).


**Reasoning**:
Handle missing values in specified columns, convert 'date_added' to datetime and extract year and month, and process the 'duration' column based on the 'type'.



In [11]:
# Example usage:
movie_title = 'Ganglands'  # Replace with a movie title from the dataset
recommendations = generate_recommendations(movie_title)

if isinstance(recommendations, str):
    print(recommendations)
else:
    print(f"Recommendations for '{movie_title}':")
    display(recommendations)

Recommendations for 'Ganglands':


2668                               Earth and Blood
3976                        The Eagle of El-Se'eed
3789                                Killer Ratings
7017    How to Live Mortgage Free with Sarah Beeny
5822                                       Cocaine
4399                                       Warrior
5210                        The Day I Met El Chapo
3297                                Paradise Beach
310            Cocaine Cowboys: The Kings of Miami
3425                                   Street Flow
Name: title, dtype: object

In [3]:
# 1. Fill missing values
data['director'].fillna('Unknown', inplace=True)
data['cast'].fillna('Unknown', inplace=True)
data['country'].fillna('Unknown', inplace=True)
data['date_added'].fillna('Not Available', inplace=True)
data['rating'].fillna('Unknown', inplace=True)

# 2. Fill missing values in 'duration' based on 'type'
data.loc[(data['type'] == 'Movie') & (data['duration'].isnull()), 'duration'] = 'Unknown Duration'
data.loc[(data['type'] == 'TV Show') & (data['duration'].isnull()), 'duration'] = 'Unknown Seasons'

# 3. Convert 'date_added' to datetime object, handling 'Not Available'
data['date_added'] = pd.to_datetime(data['date_added'], errors='coerce')

# 4. Extract year and month from 'date_added'
data['year_added'] = data['date_added'].dt.year
data['month_added'] = data['date_added'].dt.month

# Fill NaN values created by coercion in 'date_added', 'year_added', 'month_added'
data['date_added'].fillna(pd.Timestamp('1900-01-01'), inplace=True) # Using a placeholder date
data['year_added'].fillna(1900, inplace=True)
data['month_added'].fillna(1, inplace=True)

# Convert year_added and month_added to integer type after filling NaNs
data['year_added'] = data['year_added'].astype(int)
data['month_added'] = data['month_added'].astype(int)


# 5. Process 'duration'
# For Movies
movie_duration = data[data['type'] == 'Movie'].copy()
movie_duration['duration'] = movie_duration['duration'].astype(str).str.replace(' min', '', regex=False)
movie_duration['duration'] = pd.to_numeric(movie_duration['duration'], errors='coerce')
movie_duration['duration'].fillna(movie_duration['duration'].median(), inplace=True) # Fill with median

# For TV Shows
tv_show_duration = data[data['type'] == 'TV Show'].copy()
tv_show_duration['duration'] = tv_show_duration['duration'].astype(str).str.replace(' Seasons', '', regex=False).str.replace(' Season', '', regex=False)
tv_show_duration['duration'] = pd.to_numeric(tv_show_duration['duration'], errors='coerce')
tv_show_duration['duration'].fillna(tv_show_duration['duration'].median(), inplace=True) # Fill with median


# Update the original dataframe with processed duration data
data.loc[data['type'] == 'Movie', 'duration'] = movie_duration['duration']
data.loc[data['type'] == 'TV Show', 'duration'] = tv_show_duration['duration']

# Convert duration to integer type after filling NaNs
data['duration'] = data['duration'].astype(int)

display(data.head())
display(data.isnull().sum())
display(data.info())

/tmp/ipython-input-1214084273.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['director'].fillna('Unknown', inplace=True)
/tmp/ipython-input-1214084273.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try us

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water          Unknown   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans          Unknown   
4      s5  TV Show           Kota Factory          Unknown   

                                                cast        country  \
0                                            Unknown  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...        Unknown   
3                                            Unknown        Unknown   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating  duration  \
0 2021-09-25          2020  PG-13        90   
1 2021-09-24          2021  TV-MA         2   
2 2021-09-24          2021  TV-MA         1   
3 2021-09-24          2021  TV-MA         1   
4 2021-09-24          2021  TV-MA         2   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  year_added  month_added  
0  As her father nears the end of his life, filmm...        2021            9  
1  After crossing paths at a party, a Cape Town t...        2021            9  
2  To protect his family from a powerful drug lor...        2021            9  
3  Feuds, flirtations and toilet talk go down amo...        2021            9  
4  In a city of coaching centers known to train I...        2021            9

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
year_added      0
month_added     0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   object        
 1   type          8807 non-null   object        
 2   title         8807 non-null   object        
 3   director      8807 non-null   object        
 4   cast          8807 non-null   object        
 5   country       8807 non-null   object        
 6   date_added    8807 non-null   datetime64[ns]
 7   release_year  8807 non-null   int64         
 8   rating        8807 non-null   object        
 9   duration      8807 non-null   int64         
 10  listed_in     8807 non-null   object        
 11  description   8807 non-null   object        
 12  year_added    8807 non-null   int64         
 13  month_added   8807 non-null   int64         
dtypes: datetime64[ns](1), int64(4), object(9)
memory usage: 963.4+ KB


None

## Content-based filtering

### Subtask:
Implement content-based filtering by using features like genre, director, cast, and description to recommend similar movies. This might involve techniques like TF-IDF vectorization and cosine similarity.


**Reasoning**:
Combine the relevant text features for content-based filtering into a single column and display the first few rows to verify the result.



In [4]:
data['combined_features'] = data['director'] + ' ' + data['cast'] + ' ' + data['listed_in'] + ' ' + data['description']
display(data[['title', 'combined_features']].head())

title                                  combined_features
0   Dick Johnson Is Dead  Kirsten Johnson Unknown Documentaries As her f...
1          Blood & Water  Unknown Ama Qamata, Khosi Ngema, Gail Mabalane...
2              Ganglands  Julien Leclercq Sami Bouajila, Tracy Gotoas, S...
3  Jailbirds New Orleans  Unknown Unknown Docuseries, Reality TV Feuds, ...
4           Kota Factory  Unknown Mayur More, Jitendra Kumar, Ranjan Raj...

**Reasoning**:
Initialize and fit a TF-IDF Vectorizer on the combined text data and then compute the cosine similarity matrix.



In [5]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

tfidf_vectorizer = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf_vectorizer.fit_transform(data['combined_features'])

cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)
display(cosine_sim.shape)

(8807, 8807)

## Collaborative filtering

### Subtask:
Implement collaborative filtering techniques (e.g., matrix factorization or neighborhood-based methods) to find users with similar preferences and recommend movies they liked.


**Reasoning**:
I need to check if there is explicit user-item interaction data or implicit feedback in the dataset to determine the appropriate collaborative filtering technique.



In [6]:
display(data.columns)
# Check if there are columns related to user ID, item ID, or ratings/interactions.
user_interaction_columns = ['user_id', 'item_id', 'rating', 'views'] # Example column names
found_interaction_columns = [col for col in user_interaction_columns if col in data.columns]
print(f"Columns related to user interaction found: {found_interaction_columns}")

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description',
       'year_added', 'month_added', 'combined_features'],
      dtype='object')

Columns related to user interaction found: ['rating']


## Hybrid filtering

### Subtask:
Combine the results from content-based and collaborative filtering to create a hybrid recommendation system. This could involve weighted averaging, stacking, or other combination techniques.


## Hybrid filtering

### Subtask:
Combine the results from content-based and collaborative filtering to create a hybrid recommendation system. This could involve weighted averaging, stacking, or other combination techniques.


**Reasoning**:
Define the function `get_recommendations` to provide content-based recommendations based on the cosine similarity matrix.



In [7]:
def get_recommendations(title, cosine_sim=cosine_sim):
    """
    Gets content-based recommendations for a given movie title.

    Args:
        title (str): The title of the movie for which to get recommendations.
        cosine_sim (numpy.ndarray): The cosine similarity matrix.

    Returns:
        pandas.Series: A Series of the top 10 most similar movie titles.
    """
    # Get the index of the movie that matches the title
    indices = pd.Series(data.index, index=data['title']).drop_duplicates()
    if title not in indices:
        return "Movie title not found in the dataset."

    idx = indices[title]

    # Get the pairwise similarity scores of all movies with that movie
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar movies (excluding the input movie itself)
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the titles of the top 10 most similar movies
    return data['title'].iloc[movie_indices]

# Example usage:
# print(get_recommendations('Kota Factory'))

## Model evaluation

### Subtask:
Evaluate the performance of the recommendation system using appropriate metrics (e.g., precision, recall, RMSE).


**Reasoning**:
Acknowledge the limitations of the dataset and the implemented content-based approach for quantitative evaluation using standard metrics.



In [8]:
print("Due to the nature of the dataset, which lacks explicit user interaction data (like individual user ratings or viewing history), traditional collaborative filtering metrics such as RMSE, Precision, and Recall, which require ground truth user ratings or interactions, cannot be directly applied to evaluate the performance of the implemented content-based recommendation system.")
print("Evaluating a content-based recommendation system without explicit user feedback is inherently challenging.")
print("A meaningful assessment would typically require a qualitative evaluation (e.g., manually checking the relevance of recommendations for a few items) or conducting user studies to gather feedback, but these methods are outside the scope of this automated process.")
print("Therefore, a rigorous quantitative evaluation using standard recommendation system metrics is not feasible with the current data and the content-based approach alone.")

Due to the nature of the dataset, which lacks explicit user interaction data (like individual user ratings or viewing history), traditional collaborative filtering metrics such as RMSE, Precision, and Recall, which require ground truth user ratings or interactions, cannot be directly applied to evaluate the performance of the implemented content-based recommendation system.
Evaluating a content-based recommendation system without explicit user feedback is inherently challenging.
A meaningful assessment would typically require a qualitative evaluation (e.g., manually checking the relevance of recommendations for a few items) or conducting user studies to gather feedback, but these methods are outside the scope of this automated process.
Therefore, a rigorous quantitative evaluation using standard recommendation system metrics is not feasible with the current data and the content-based approach alone.


## Recommendation generation

### Subtask:
Develop a function or module to generate recommendations for a given user or movie.


**Reasoning**:
Define the `generate_recommendations` function as per the instructions, which will call the previously defined `get_recommendations` function.



In [9]:
def generate_recommendations(movie_title):
    """
    Generates content-based recommendations for a given movie title.

    Args:
        movie_title (str): The title of the movie for which to get recommendations.

    Returns:
        pandas.Series or str: A Series of the top 10 most similar movie titles
                                or a string message if the movie is not found.
    """
    return get_recommendations(movie_title)

# Example usage:
# print(generate_recommendations('Zodiac'))

## Summary:

### Data Analysis Key Findings

*   The dataset contains 8807 entries and 12 columns, with several columns having missing values (director, cast, country, date\_added, rating, and duration).
*   Missing values in various columns were handled by filling them with placeholder values like 'Unknown', 'Not Available', or median values based on content type for duration.
*   New features `year_added` and `month_added` were successfully extracted from the `date_added` column.
*   Content-based filtering was implemented by combining text features and computing a cosine similarity matrix using TF-IDF vectorization. The shape of the cosine similarity matrix is (8807, 8807).
*   Collaborative filtering could not be implemented due to the absence of explicit user interaction data (user IDs, individual ratings/views) in the dataset.
*   Due to the lack of user interaction data, traditional recommendation system evaluation metrics like RMSE, Precision, and Recall were not applicable.
*   A function `generate_recommendations` was created to provide content-based recommendations using the pre-computed cosine similarity matrix.

### Insights or Next Steps

*   The current system is solely content-based due to data limitations. To build a truly hybrid system, acquiring or simulating user interaction data would be necessary.
*   Evaluating the content-based recommendations would require qualitative methods or user studies to assess the relevance of the suggested titles.


In [21]:
!streamlit run app.py & npx localtunnel --port 8501

⠙

⠹⠸⠼⠴⠦
  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://172.28.0.12:8501
  External URL: http://35.229.165.4:8501

⠧⠇⠏⠋⠙⠹⠸your url is: https://stupid-rabbits-play.loca.lt
/content/app.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['director'].fillna('Unknown', inplace=True)
/content/app.py:12: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pa

# Task
Create a Streamlit application for a hybrid movie recommendation system using the dataset "/content/sample_data/netflix_titles.csv".

## Install streamlit

### Subtask:
Install the Streamlit library in your environment.


**Reasoning**:
Install the streamlit library using pip.



In [12]:
!pip install streamlit

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.1/10.1 MB 52.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/6.9 MB 90.2 MB/s eta 0:00:00


## Create a streamlit app file

### Subtask:
Create a new Python file that will contain the code for your Streamlit application.


**Reasoning**:
Create an empty file named `app.py`.



In [13]:
!touch app.py

**Reasoning**:
Verify that the file `app.py` was created successfully.



In [14]:
!ls

app.py	sample_data


## Import libraries and load data

### Subtask:
Import necessary libraries (pandas, sklearn, streamlit) and load the Netflix data. Reuse the data loading and preprocessing steps from the notebook.


**Reasoning**:
Import necessary libraries, load the data, and apply the preprocessing steps from the notebook, ensuring inplace arguments are handled correctly and duration is processed accurately.



In [15]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import streamlit as st

# Load the dataset
data = pd.read_csv('/content/sample_data/netflix_titles.csv')

# Preprocessing steps from the notebook
# 1. Fill missing values
data['director'] = data['director'].fillna('Unknown')
data['cast'] = data['cast'].fillna('Unknown')
data['country'] = data['country'].fillna('Unknown')
data['date_added'] = data['date_added'].fillna('Not Available')
data['rating'] = data['rating'].fillna('Unknown')

# 2. Fill missing values in 'duration' based on 'type'
data.loc[(data['type'] == 'Movie') & (data['duration'].isnull()), 'duration'] = 'Unknown Duration'
data.loc[(data['type'] == 'TV Show') & (data['duration'].isnull()), 'duration'] = 'Unknown Seasons'

# 3. Convert 'date_added' to datetime object, handling 'Not Available'
data['date_added'] = pd.to_datetime(data['date_added'], errors='coerce')

# 4. Extract year and month from 'date_added'
data['year_added'] = data['date_added'].dt.year
data['month_added'] = data['date_added'].dt.month

# Fill NaN values created by coercion in 'date_added', 'year_added', 'month_added'
data['date_added'] = data['date_added'].fillna(pd.Timestamp('1900-01-01')) # Using a placeholder date
data['year_added'] = data['year_added'].fillna(1900).astype(int)
data['month_added'] = data['month_added'].fillna(1).astype(int)

# 5. Process 'duration'
# For Movies
movie_duration_mask = data['type'] == 'Movie'
data.loc[movie_duration_mask, 'duration'] = data.loc[movie_duration_mask, 'duration'].astype(str).str.replace(' min', '', regex=False)
data.loc[movie_duration_mask, 'duration'] = pd.to_numeric(data.loc[movie_duration_mask, 'duration'], errors='coerce')
movie_median_duration = data.loc[movie_duration_mask, 'duration'].median()
data.loc[movie_duration_mask, 'duration'] = data.loc[movie_duration_mask, 'duration'].fillna(movie_median_duration)

# For TV Shows
tv_show_duration_mask = data['type'] == 'TV Show'
data.loc[tv_show_duration_mask, 'duration'] = data.loc[tv_show_duration_mask, 'duration'].astype(str).str.replace(' Seasons', '', regex=False).str.replace(' Season', '', regex=False)
data.loc[tv_show_duration_mask, 'duration'] = pd.to_numeric(data.loc[tv_show_duration_mask, 'duration'], errors='coerce')
tv_show_median_duration = data.loc[tv_show_duration_mask, 'duration'].median()
data.loc[tv_show_duration_mask, 'duration'] = data.loc[tv_show_duration_mask, 'duration'].fillna(tv_show_median_duration)

# Convert duration to integer type after filling NaNs
data['duration'] = data['duration'].astype(int)

# Combine features for content-based filtering
data['combined_features'] = data['director'] + ' ' + data['cast'] + ' ' + data['listed_in'] + ' ' + data['description']

# Display the first few rows and check for null values
display(data.head())
display(data.isnull().sum())

/tmp/ipython-input-2243047444.py:39: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data.loc[movie_duration_mask, 'duration'] = data.loc[movie_duration_mask, 'duration'].fillna(movie_median_duration)
/tmp/ipython-input-2243047444.py:46: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data.loc[tv_show_duration_mask, 'duration'] = data.loc[tv_show_duration_mask, 'duration'].fillna(tv_show_median_duration)


show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water          Unknown   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans          Unknown   
4      s5  TV Show           Kota Factory          Unknown   

                                                cast        country  \
0                                            Unknown  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...        Unknown   
3                                            Unknown        Unknown   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating  duration  \
0 2021-09-25          2020  PG-13        90   
1 2021-09-24          2021  TV-MA         2   
2 2021-09-24          2021  TV-MA         1   
3 2021-09-24          2021  TV-MA         1   
4 2021-09-24          2021  TV-MA         2   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  year_added  month_added  \
0  As her father nears the end of his life, filmm...        2021            9   
1  After crossing paths at a party, a Cape Town t...        2021            9   
2  To protect his family from a powerful drug lor...        2021            9   
3  Feuds, flirtations and toilet talk go down amo...        2021            9   
4  In a city of coaching centers known to train I...        2021            9   

                                   combined_features  
0  Kirsten Johnson Unknown Documentaries As her f...  
1  Unknown Ama Qamata, Khosi Ngema, Gail Mabalane...  
2  Julien Leclercq Sami Bouajila, Tracy Gotoas, S...  
3  Unknown Unknown Docuseries, Reality TV Feuds, ...  
4  Unknown Mayur More, Jitendra Kumar, Ranjan Raj...

show_id              0
type                 0
title                0
director             0
cast                 0
country              0
date_added           0
release_year         0
rating               0
duration             0
listed_in            0
description          0
year_added           0
month_added          0
combined_features    0
dtype: int64

**Reasoning**:
The data loading and preprocessing steps are complete and verified. The next step is to implement the content-based filtering logic by creating the TF-IDF vectorizer and computing the cosine similarity matrix.



In [16]:
# Create TF-IDF vectorizer and fit on combined features
tfidf_vectorizer = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf_vectorizer.fit_transform(data['combined_features'])

# Compute cosine similarity matrix
cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

# Display the shape of the cosine similarity matrix
display(cosine_sim.shape)

(8807, 8807)

## Implement recommendation logic

### Subtask:
Copy the relevant code for the content-based filtering, including the TF-IDF vectorizer, cosine similarity calculation, and the `get_recommendations` function, into the Streamlit app file.


**Reasoning**:
Copy the content-based filtering code into the Streamlit app file.



In [17]:
%%writefile app.py
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import streamlit as st

# Load the data
data = pd.read_csv('/content/sample_data/netflix_titles.csv')

# Data preprocessing
data['director'].fillna('Unknown', inplace=True)
data['cast'].fillna('Unknown', inplace=True)
data['country'].fillna('Unknown', inplace=True)
data['date_added'].fillna('Not Available', inplace=True)
data['rating'].fillna('Unknown', inplace=True)

data.loc[(data['type'] == 'Movie') & (data['duration'].isnull()), 'duration'] = 'Unknown Duration'
data.loc[(data['type'] == 'TV Show') & (data['duration'].isnull()), 'duration'] = 'Unknown Seasons'

data['date_added'] = pd.to_datetime(data['date_added'], errors='coerce')
data['year_added'] = data['date_added'].dt.year
data['month_added'] = data['date_added'].dt.month

data['date_added'].fillna(pd.Timestamp('1900-01-01'), inplace=True)
data['year_added'].fillna(1900, inplace=True)
data['month_added'].fillna(1, inplace=True)

data['year_added'] = data['year_added'].astype(int)
data['month_added'] = data['month_added'].astype(int)

movie_duration = data[data['type'] == 'Movie'].copy()
movie_duration['duration'] = movie_duration['duration'].astype(str).str.replace(' min', '', regex=False)
movie_duration['duration'] = pd.to_numeric(movie_duration['duration'], errors='coerce')
movie_duration['duration'].fillna(movie_duration['duration'].median(), inplace=True)

tv_show_duration = data[data['type'] == 'TV Show'].copy()
tv_show_duration['duration'] = tv_show_duration['duration'].astype(str).str.replace(' Seasons', '', regex=False).str.replace(' Season', '', regex=False)
tv_show_duration['duration'] = pd.to_numeric(tv_show_duration['duration'], errors='coerce')
tv_show_duration['duration'].fillna(tv_show_duration['duration'].median(), inplace=True)

data.loc[data['type'] == 'Movie', 'duration'] = movie_duration['duration']
data.loc[data['type'] == 'TV Show', 'duration'] = tv_show_duration['duration']
data['duration'] = data['duration'].astype(int)

# Content-based filtering code
data['combined_features'] = data['director'] + ' ' + data['cast'] + ' ' + data['listed_in'] + ' ' + data['description']

tfidf_vectorizer = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf_vectorizer.fit_transform(data['combined_features'])

cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

def get_recommendations(title, cosine_sim=cosine_sim):
    """
    Gets content-based recommendations for a given movie title.

    Args:
        title (str): The title of the movie for which to get recommendations.
        cosine_sim (numpy.ndarray): The cosine similarity matrix.

    Returns:
        pandas.Series: A Series of the top 10 most similar movie titles.
    """
    indices = pd.Series(data.index, index=data['title']).drop_duplicates()
    if title not in indices.index: # Use indices.index to check if the title exists
        return "Movie title not found in the dataset."

    idx = indices[title]

    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:11]

    movie_indices = [i[0] for i in sim_scores]
    return data['title'].iloc[movie_indices]

# Streamlit app interface (initial structure)
st.title('Netflix Movie Recommendation System')

Overwriting app.py


## Design streamlit interface

### Subtask:
Use Streamlit components (like `st.title`, `st.selectbox`, `st.button`, `st.write`) to create a user interface where the user can select a movie and get recommendations.


**Reasoning**:
Add Streamlit components to create the user interface for selecting a movie and displaying recommendations.



In [18]:
%%writefile app.py
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import streamlit as st

# Load the data
data = pd.read_csv('/content/sample_data/netflix_titles.csv')

# Data preprocessing
data['director'].fillna('Unknown', inplace=True)
data['cast'].fillna('Unknown', inplace=True)
data['country'].fillna('Unknown', inplace=True)
data['date_added'].fillna('Not Available', inplace=True)
data['rating'].fillna('Unknown', inplace=True)

data.loc[(data['type'] == 'Movie') & (data['duration'].isnull()), 'duration'] = 'Unknown Duration'
data.loc[(data['type'] == 'TV Show') & (data['duration'].isnull()), 'duration'] = 'Unknown Seasons'

data['date_added'] = pd.to_datetime(data['date_added'], errors='coerce')
data['year_added'] = data['date_added'].dt.year
data['month_added'] = data['date_added'].dt.month

data['date_added'].fillna(pd.Timestamp('1900-01-01'), inplace=True)
data['year_added'].fillna(1900, inplace=True)
data['month_added'].fillna(1, inplace=True)

data['year_added'] = data['year_added'].astype(int)
data['month_added'] = data['month_added'].astype(int)

movie_duration = data[data['type'] == 'Movie'].copy()
movie_duration['duration'] = movie_duration['duration'].astype(str).str.replace(' min', '', regex=False)
movie_duration['duration'] = pd.to_numeric(movie_duration['duration'], errors='coerce')
movie_duration['duration'].fillna(movie_duration['duration'].median(), inplace=True)

tv_show_duration = data[data['type'] == 'TV Show'].copy()
tv_show_duration['duration'] = tv_show_duration['duration'].astype(str).str.replace(' Seasons', '', regex=False).str.replace(' Season', '', regex=False)
tv_show_duration['duration'] = pd.to_numeric(tv_show_duration['duration'], errors='coerce')
tv_show_duration['duration'].fillna(tv_show_duration['duration'].median(), inplace=True)

data.loc[data['type'] == 'Movie', 'duration'] = movie_duration['duration']
data.loc[data['type'] == 'TV Show', 'duration'] = tv_show_duration['duration']
data['duration'] = data['duration'].astype(int)


# Content-based filtering code
data['combined_features'] = data['director'] + ' ' + data['cast'] + ' ' + data['listed_in'] + ' ' + data['description']

tfidf_vectorizer = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf_vectorizer.fit_transform(data['combined_features'])

cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix)

def get_recommendations(title, cosine_sim=cosine_sim):
    """
    Gets content-based recommendations for a given movie title.

    Args:
        title (str): The title of the movie for which to get recommendations.
        cosine_sim (numpy.ndarray): The cosine similarity matrix.

    Returns:
        pandas.Series or str: A Series of the top 10 most similar movie titles
                                or a string message if the movie is not found.
    """
    indices = pd.Series(data.index, index=data['title']).drop_duplicates()
    if title not in indices.index:
        return "Movie title not found in the dataset."

    idx = indices[title]

    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:11]

    movie_indices = [i[0] for i in sim_scores]
    return data['title'].iloc[movie_indices]

# Streamlit app interface
st.title('Netflix Movie Recommendation System')

# Selectbox for movie title
selected_movie = st.selectbox('Select a movie:', data['title'])

# Button to trigger recommendations
if st.button('Get Recommendations'):
    recommendations = get_recommendations(selected_movie)

    if isinstance(recommendations, str):
        st.write(recommendations)
    else:
        st.write(f"Recommendations for '{selected_movie}':")
        st.dataframe(recommendations)

Overwriting app.py


## Run the streamlit app

### Subtask:
Provide instructions on how to run the Streamlit application.


## Summary:

### Data Analysis Key Findings

*   The `streamlit` library was successfully installed using `pip install streamlit`.
*   An empty Python file named `app.py` was created using `!touch app.py` to house the Streamlit application code.
*   The necessary libraries (pandas, numpy, sklearn, streamlit) were imported, and the Netflix dataset was loaded and preprocessed within the `app.py` file.
*   Missing values in key columns ('director', 'cast', 'country', 'date\_added', 'rating', 'duration') were handled using strategies like filling with 'Unknown', 'Not Available', median values, or placeholders.
*   New features ('year\_added', 'month\_added', 'combined\_features') were created for the recommendation system.
*   A TF-IDF vectorizer was fitted to the `combined_features`, and the cosine similarity matrix was computed, resulting in a shape of 8807x8807.
*   The `get_recommendations` function, which uses the cosine similarity matrix to find similar movie titles, was implemented in `app.py`.
*   Streamlit components, including `st.title`, `st.selectbox` for movie selection, `st.button` to trigger recommendations, and `st.write`/`st.dataframe` to display results or error messages, were added to the `app.py` file to create the user interface.
*   The final `app.py` file contains the complete code for loading data, preprocessing, implementing the content-based recommendation logic, and designing the Streamlit interface.

### Insights or Next Steps

*   The current application is a content-based recommendation system. A next step could be to integrate collaborative filtering or other methods to build a hybrid system as initially intended, potentially improving recommendation quality.
*   To improve the application's performance and user experience, the TF-IDF matrix and cosine similarity matrix could be pre-computed and saved, then loaded by the Streamlit app, rather than recalculating them every time the app runs.


# Task
Create a movie recommendation system using hybrid filtering with the dataset "/content/sample_data/netflix_titles.csv", and build a Streamlit application for it.

## Pre-compute and save matrices

### Subtask:
Modify the notebook to compute the TF-IDF matrix and cosine similarity matrix and save them to files.


**Reasoning**:
Compute the TF-IDF matrix and cosine similarity matrix and save them to files using pickle.



In [22]:
import pickle

# Compute the TF-IDF matrix and cosine similarity matrix (using existing code)
# tfidf_vectorizer = TfidfVectorizer(stop_words='english') # Already defined and fitted
# tfidf_matrix = tfidf_vectorizer.fit_transform(data['combined_features']) # Already computed
# cosine_sim = cosine_similarity(tfidf_matrix, tfidf_matrix) # Already computed

# Save the TF-IDF matrix
with open('tfidf_matrix.pkl', 'wb') as f:
    pickle.dump(tfidf_matrix, f)

# Save the cosine similarity matrix
with open('cosine_sim.pkl', 'wb') as f:
    pickle.dump(cosine_sim, f)

print("TF-IDF matrix and cosine similarity matrix saved to tfidf_matrix.pkl and cosine_sim.pkl")

TF-IDF matrix and cosine similarity matrix saved to tfidf_matrix.pkl and cosine_sim.pkl


## Update streamlit app to load matrices

### Subtask:
Modify the `app.py` file to load the pre-computed matrices instead of recalculating them.


**Reasoning**:
Modify the app.py file to load the pre-computed matrices instead of recalculating them.



In [23]:
%%writefile app.py
import pandas as pd
import numpy as np
import pickle
import streamlit as st

# Load the data
data = pd.read_csv('/content/sample_data/netflix_titles.csv')

# Data preprocessing
data['director'].fillna('Unknown', inplace=True)
data['cast'].fillna('Unknown', inplace=True)
data['country'].fillna('Unknown', inplace=True)
data['date_added'].fillna('Not Available', inplace=True)
data['rating'].fillna('Unknown', inplace=True)

data.loc[(data['type'] == 'Movie') & (data['duration'].isnull()), 'duration'] = 'Unknown Duration'
data.loc[(data['type'] == 'TV Show') & (data['duration'].isnull()), 'duration'] = 'Unknown Seasons'

data['date_added'] = pd.to_datetime(data['date_added'], errors='coerce')
data['year_added'] = data['date_added'].dt.year
data['month_added'] = data['date_added'].dt.month

data['date_added'].fillna(pd.Timestamp('1900-01-01'), inplace=True)
data['year_added'].fillna(1900, inplace=True)
data['month_added'].fillna(1, inplace=True)

data['year_added'] = data['year_added'].astype(int)
data['month_added'] = data['month_added'].astype(int)

movie_duration = data[data['type'] == 'Movie'].copy()
movie_duration['duration'] = movie_duration['duration'].astype(str).str.replace(' min', '', regex=False)
movie_duration['duration'] = pd.to_numeric(movie_duration['duration'], errors='coerce')
movie_duration['duration'].fillna(movie_duration['duration'].median(), inplace=True)

tv_show_duration = data[data['type'] == 'TV Show'].copy()
tv_show_duration['duration'] = tv_show_duration['duration'].astype(str).str.replace(' Seasons', '', regex=False).str.replace(' Season', '', regex=False)
tv_show_duration['duration'] = pd.to_numeric(tv_show_duration['duration'], errors='coerce')
tv_show_duration['duration'].fillna(tv_show_duration['duration'].median(), inplace=True)

data.loc[data['type'] == 'Movie', 'duration'] = movie_duration['duration']
data.loc[data['type'] == 'TV Show', 'duration'] = tv_show_duration['duration']
data['duration'] = data['duration'].astype(int)

# Load pre-computed matrices
try:
    with open('/content/tfidf_matrix.pkl', 'rb') as f:
        tfidf_matrix = pickle.load(f)
    with open('/content/cosine_sim.pkl', 'rb') as f:
        cosine_sim = pickle.load(f)
except FileNotFoundError:
    st.error("Pre-computed matrices not found. Please run the notebook cells to generate them.")
    st.stop() # Stop the app if files are not found


def get_recommendations(title, cosine_sim=cosine_sim):
    """
    Gets content-based recommendations for a given movie title.

    Args:
        title (str): The title of the movie for which to get recommendations.
        cosine_sim (numpy.ndarray): The cosine similarity matrix.

    Returns:
        pandas.Series or str: A Series of the top 10 most similar movie titles
                                or a string message if the movie is not found.
    """
    indices = pd.Series(data.index, index=data['title']).drop_duplicates()
    if title not in indices.index:
        return "Movie title not found in the dataset."

    idx = indices[title]

    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:11]

    movie_indices = [i[0] for i in sim_scores]
    return data['title'].iloc[movie_indices]

# Streamlit app interface
st.title('Netflix Movie Recommendation System')

# Selectbox for movie title
selected_movie = st.selectbox('Select a movie:', data['title'])

# Button to trigger recommendations
if st.button('Get Recommendations'):
    recommendations = get_recommendations(selected_movie)

    if isinstance(recommendations, str):
        st.write(recommendations)
    else:
        st.write(f"Recommendations for '{selected_movie}':")
        st.dataframe(recommendations)

Overwriting app.py


## Run the optimized streamlit app

### Subtask:
Run the updated Streamlit application to see the improved startup time.


**Reasoning**:
Run the Streamlit application with localtunnel to make it accessible.



In [24]:
!streamlit run app.py & npx localtunnel --port 8501

⠙

⠹⠸⠼⠴
  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://172.28.0.12:8501
  External URL: http://35.229.165.4:8501

⠦⠧⠇⠏⠋⠙⠹⠸your url is: https://silver-papayas-marry.loca.lt
/content/app.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['director'].fillna('Unknown', inplace=True)
/content/app.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in p

In [25]:
%%writefile requirements.txt
streamlit
pandas
numpy
scikit-learn

Writing requirements.txt
